src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,147 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final int START_SIZE = 0;
+    private static final int INDEX_LOWER_BOUND = 0;
+    private static final double CAPACITY_MULTIPLIER = 1.5;
+    private static final String INVALID_INDEX_MESSAGE = "The provided index is invalid. "
+            + "Please double-check it. Index: ";
+    private Object[] elementData;
+    private int size;
+
+    public ArrayList() {
+        elementData = new Object[DEFAULT_CAPACITY];
+        size = START_SIZE;
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == elementData.length) {
+            elementData = grow();
+        }
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        rangeCheckForAdd(index);
+        if (size == elementData.length) {
+            elementData = grow();
+        }
+        System.arraycopy(elementData, index, elementData,
+                index + 1, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (size + list.size() >= elementData.length) {
+            elementData = grow();
+        }
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        rangeCheckForRemoveAndSetAndGet(index);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        rangeCheckForRemoveAndSetAndGet(index);
+        if (size == elementData.length) {
```

Comment:
I should delete this check here in case that set method doesn't push new values to array but only changes existing.

Reply:
that's a good point, thanks